Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/.gitignore

```diff
@@ -0,0 +1,33 @@
+# See https://help.github.com/articles/ignoring-files for more about ignoring files.
```

Comment:
what is this file for?

Reply:
`.gitignore` is used to list out files that we want to be excluded during git push. This can be useful to exclude some local files that includes private information such as `.env` file or some unrelated files that get pushed accidentally to the repository such as `.DS_Store` in Mac.

Reply:
```
some unrelated files that get pushed accidentally to the repository such as .DS_Store in Mac
```

Can you try this statement? Push a file by accident to github. Then add that file path to here. See if its ignored or not.

Reply:
What I mean is that using `.gitignore` could prevent accidentally pushing unwanted files to the repository, but of course you have to include the path of the unwanted files inside the `.gitignore` file beforehand.

Reply:
2. To push a file by accident to Github, I created a file named `unwantedfile` inside the `testing_project` directory then pushing it to git before adding in `unwantedfile` inside the `.gitignore` file.

In order to make the repository to be ignoring `unwantedfile` correctly,
run:
```
git rm --cached unwantedfile
```
then `git commit` with the correct message before `git push`,  your `.gitignore` will now works correctly.

Reply:
very good.

Now, try one more thing. Look at your `git log`. You will have a log to show you had committed that unwanted file and you can see the record of that file. How do you remove the history of that unwanted file?

Reply:
To remove the commit history from the git log, 
first run:
```
git log --oneline
```
to show all history in one line along with the 7-digits sha1-commit-id.

In this case, since I already removed the `unwantedfile` and all I wanted is to remove the commit history with commits message starting with `test` without any further changes to my working directory and staging area, so I use `git rebase` by running:
```
git rebase -i <sha-commit-id>
```
with `<sha-commit-id>` is the sha commit id of the previous commit and `-i` command opens up the interactive mode so I can remove the two commit with `test` in the editor.

Lastly, I run:
```
git push --force-with-lease origin testing-initialization
```
to force the push and making remote branch up to date with the local branch.